doc/man/fedfs-get-nsdb-params.8
.\"@(#)fedfs-get-nsdb-params.8"
.\"
.\" @file doc/man/fedfs-get-nsdb-params.8
.\" @brief man page for fedfs-get-nsdb-params client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH FEDFS-GET-NSDB-PARAMS 8 "@publication-date@"
.SH NAME
fedfs-get-nsdb-params \- send a FEDFS_GET_NSDB_PARAMS ADMIN protocol request
.SH SYNOPSIS
.B fedfs-get-nsdb-params
.RB [ \-?d ]
.RB [ \-f
.IR certfile ]
.RB [ \-n
.IR nettype ]
.RB [ \-h
.IR hostname ]
.RB [ \-l
.IR nsdbname ]
.RB [ \-r
.IR nsdbport ]
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
FedFS-enabled file servers allow remote administrative access via an
authenticated RPC protocol known as the
.IR "FedFS ADMIN protocol" .
Using this protocol, FedFS administrators manage
FedFS junctions and NSDB connection parameter information
on remote FedFS-enabled file servers.
.SH DESCRIPTION
The
.BR fedfs-get-nsdb-params (8)
command is part of a collection of low-level single-use programs
that is intended for testing the FedFS ADMIN protocol or for use in scripts.
It sends a single FEDFS_GET_NSDB_PARAMS request to a remote
FedFS ADMIN protocol service.
.P
The FEDFS_GET_NSDB_PARAMS request retrieves
NSDB connection parameter information stored on a remote server.
For more on the specification and use of NSDB connection parameters, see
.BR nsdbparams "(8) or"
.BR fedfs (7).
.P
An NSDB hostname and port number (see below)
are used as the primary key to identify an entry
in the remote server's NSDB connection parameter database.
Details on NSDB connection parameter database entry matching can be
found in
.BR nsdb-parameters (7).
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR fedfs-get-nsdb-params (8)
version information and a usage message on
.IR stderr .
.IP "\fB-f, \-\-certfile=\fIpathname\fP"
Specifies the pathname of a local file to write received certificate
material into, when the specified NSDB's connection security type is
.BR TLS .
If no file is specified, the certificate is ignored.
The certificate is written to the specified file in PEM format.
.IP "\fB\-h, \-\-hostname=\fIhostname\fP"
Specifies the hostname of a remote FedFS ADMIN service.
If this option is not specified, the default value is
.BR localhost .
.IP "\fB\-n, \-\-nettype=\fInettype\fP"
Specifies the transport to use when contacting the remote FedFS ADMIN service.
Typically the
.I nettype
is one of
.B tcp
or
.BR udp .
If this option is not specified, the default value is
.BR netpath .
See
.BR rpc (3t)
for details.
.IP "\fB\-l, \-\-nsdbname=\fINSDB-hostname\fP"
Specifies the hostname of the NSDB to insert into the new FedFS junction.
If this option is not specified,
the value of the FEDFS_NSDB_HOST environment variable is consulted.
If the variable is not set and the
.B \-\-nsdbname
option is not specified, the
.BR fedfs-get-nsdb-params (8)
command fails.
.IP "\fB\-r, \-\-nsdbport=\fINSDB-port\fP"
Specifies the IP port of the NSDB to insert into the new FedFS junction.
If this option is not specified,
the value of the FEDFS_NSDB_PORT environment variable is consulted.
The default value if the variable is not set is 389.
.SH EXIT CODES
The remote administration service returns a value that reflects the
success of the requested operation.
.TP
.B FEDFS_OK
The operation succeeded.
.TP
.B FEDFS_ERR_ACCESS
The caller does not have permission to perform the requested operation
on the remote server.
.TP
.B FEDFS_ERR_BADCHAR
The NSDB hostname contains a character which is not
supported by the remote server.
.TP
.B FEDFS_ERR_BADXDR
The remote server encountered an XDR decoding error while
processing the request.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_IO
A hard error occurred on the remote server.
.TP
.B FEDFS_ERR_PERM
The operation was not allowed because the caller is
either not a privileged user or not the owner of an object that
would be modified by the operation.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred on the remote server.
.TP
.B FEDFS_ERR_NSDB_ROUTE
The remote server was unable to find a route to the NSDB.
.TP
.B FEDFS_ERR_NSDB_DOWN
The remote server determined that the NSDB was down.
.TP
.B FEDFS_ERR_NSDB_CONN
The remote server was unable to establish a connection with the NSDB.
.TP
.B FEDFS_ERR_NSDB_AUTH
The remote server was unable to authenticate
and establish a secure connection with the NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP
An LDAP error occurred on the connection between the remote server and NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_VAL
Indicates the same error as FEDFS_ERR_NSDB_LDAP,
and allows an LDAP protocol error value to be returned to the client.
.TP
.B FEDFS_ERR_NSDB_RESPONSE
The remote server received a malformed response from the NSDB.
This includes situations when an NSDB entry (e.g. FSN or FSL)
is missing a required attribute.
.TP
.B FEDFS_ERR_NSDB_FAULT
An unanticipated error related to the NSDB occurred.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The remote server does not have any connection
parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL
The remote server received an LDAP referral that it was unable to follow.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_VAL
Indicates the same error as FEDFS_ERR_NSDB_LDAP_REFERRAL,
and allows an LDAP protocol error value to be returned back to the client.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_NOTFOLLOWED
The remote server received an LDAP referral that it chose not to follow,
either because the remote server does not support following LDAP referrals
or LDAP referral following is disabled.
.TP
.B FEDFS_ERR_NSDB_PARAMS_LDAP_REFERRAL
The remote server received an LDAP referral that it chose not to follow
because the remote server had no NSDB parameters for the NSDB
targeted by the LDAP referral.
.TP
.B FEDFS_ERR_NOTSUPP
The remote server does not support the specified procedure.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that your domain's NSDB hostname is
.IR nsdb.example.net .
If the file server
.IR fs.example.net
already knows about your domain's NSDB, you can query it with:
.RS
.sp
$ fedfs-get-nsdb-params -h fs.example.net -l nsdb.example.net
.br
Call completed successfully
.br
ConnectionSec: FEDFS_SEC_NONE
.sp
.RE
The remote server knows about
.I nsdb.example.net
and does not use TLS when querying it to resolve junctions.
.SH SECURITY
RPCSEC GSSAPI authentication has not yet been implemented for this command.
.SH "SEE ALSO"
.BR fedfs (7),
.BR nsdb-parameters (7),
.BR rpc.fedfsd (8),
.BR fedfs-get-nsdb-limited-params (8),
.BR nsdbparams (8),
.BR rpc (3t)
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/fedfsc/fedfs-get-nsdb-params.c
/**
 * @file src/fedfsc/fedfs-get-nsdb-params.c
 * @brief Send a FEDFS_GET_NSDB_PARAMS RPC to a FedFS ADMIN server
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <string.h>
#include <stdbool.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>

#include <rpc/clnt.h>
#include <ldap.h>

#include "fedfs.h"
#include "fedfs_admin.h"
#include "nsdb.h"
#include "junction.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Default RPC request timeout
 */
static struct timeval fedfs_get_nsdb_params_timeout = { 25, 0 };

/**
 * Short form command line options
 */
static const char fedfs_get_nsdb_params_opts[] = "?df:h:l:n:r:";

/**
 * Long form command line options
 */
static const struct option fedfs_get_nsdb_params_longopts[] = {
	{ "certfile", 1, NULL, 'f', },
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "hostname", 1, NULL, 'h', },
	{ "nsdbname", 1, NULL, 'l', },
	{ "nettype", 1, NULL, 'n', },
	{ "nsdbport", 1, NULL, 'r', },
	{ NULL, 0, NULL, 0, },
};

static void
fedfs_get_nsdb_params_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [-d] [-n nettype] [-h hostname] "
			"[-f certfile] [-l nsdbname] [-r nsdbport]\n\n",
			progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-f, --certfile       Where to write certificate\n");
	fprintf(stderr, "\t-n, --nettype        RPC transport (default: 'netpath')\n");
	fprintf(stderr, "\t-h, --hostname       ADMIN server hostname (default: 'localhost')\n");
	fprintf(stderr, "\t-l, --nsdbname       NSDB hostname\n");
	fprintf(stderr, "\t-r, --nsdbport       NSDB port\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

static void
fedfs_get_nsdb_params_print_result(FedFsGetNsdbParamsRes result,
		const char *certfile)
{
	FedFsNsdbParams *params = &result.FedFsGetNsdbParamsRes_u.params;

	nsdb_print_fedfsstatus(result.status);
	if (result.status != FEDFS_OK)
		return;

	switch (params->secType) {
	case FEDFS_SEC_NONE:
		printf("ConnectionSec: FEDFS_SEC_NONE\n");
		break;
	case FEDFS_SEC_TLS:
		printf("ConnectionSec: FEDFS_SEC_TLS\n");
		if (certfile != NULL)
			(void)nsdb_connsec_write_pem_file(certfile,
				params->FedFsNsdbParams_u.secData.secData_val,
				params->FedFsNsdbParams_u.secData.secData_len);
		break;
	default:
		printf("Unrecognized FedFsConnectionSec value: %u\n",
				params->secType);
	}
}

static FedFsStatus
fedfs_get_nsdb_params_call(const char *hostname, const char *nettype,
		char *nsdbname, const unsigned short nsdbport,
		const char *certfile)
{
	FedFsGetNsdbParamsRes result;
	enum clnt_stat status;
	FedFsNsdbName arg;
	CLIENT *client;

	memset(&arg, 0, sizeof(arg));

	arg.hostname.utf8string_len = strlen(nsdbname);
	arg.hostname.utf8string_val = nsdbname;
	arg.port = nsdbport;

	client = clnt_create(hostname, FEDFS_PROG, FEDFS_V1, nettype);
	if (client == NULL) {
		clnt_pcreateerror("Failed to create FEDFS client");
		result.status = FEDFS_ERR_SVRFAULT;
		goto out;
	}

	memset((char *)&result, 0, sizeof(result));
	status = clnt_call(client, FEDFS_GET_NSDB_PARAMS,
				(xdrproc_t)xdr_FedFsNsdbName, (caddr_t)&arg,
				(xdrproc_t)xdr_FedFsGetNsdbParamsRes, (caddr_t)&result,
				fedfs_get_nsdb_params_timeout);
	if (status != RPC_SUCCESS) {
		clnt_perror(client, "FEDFS_GET_NSDB_PARAMS call failed");
		result.status = FEDFS_ERR_SVRFAULT;
	} else {
		fedfs_get_nsdb_params_print_result(result, certfile);
		clnt_freeres(client,
			(xdrproc_t)xdr_FedFsGetNsdbParamsRes,
			(caddr_t)&result);
	}
	(void)clnt_destroy(client);

out:
	return result.status;
}

int
main(int argc, char **argv)
{
	char *progname, *hostname, *nettype, *nsdbname, *certfile;
	unsigned short nsdbport;
	unsigned int seconds;
	FedFsStatus status;
	int arg;

	(void)setlocale(LC_ALL, "");
	(void)umask(S_IRWXO);

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	nsdb_env(&nsdbname, &nsdbport, NULL, NULL);

	hostname = "localhost";
	nettype = "netpath";
	certfile = NULL;
	while ((arg = getopt_long(argc, argv, fedfs_get_nsdb_params_opts, fedfs_get_nsdb_params_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'f':
			certfile = optarg;
			break;
		case 'h':
			hostname = optarg;
			break;
		case 'l':
			if (!nsdb_is_hostname_utf8(optarg)) {
				fprintf(stderr, "NSDB name %s is "
					"not a UTF-8 hostname\n", optarg);
				fedfs_get_nsdb_params_usage(progname);
			}
			nsdbname = optarg;
			break;
		case 'n':
			nettype = optarg;
			break;
		case 'r':
			if (!nsdb_parse_port_string(optarg, &nsdbport)) {
				fprintf(stderr, "Bad port number: %s\n",
					optarg);
				fedfs_get_nsdb_params_usage(progname);
			}
			break;
		default:
			fprintf(stderr, "Invalid command line argument: %c\n", (char)arg);
		case '?':
			fedfs_get_nsdb_params_usage(progname);
		}
	}
	if (optind != argc) {
		fprintf(stderr, "Unrecognized command line argument\n");
		fedfs_get_nsdb_params_usage(progname);
	}
	if (nsdbname == NULL) {
		fprintf(stderr, "Missing required command line argument\n");
		fedfs_get_nsdb_params_usage(progname);
	}

	nsdb_connsec_crypto_startup();

	for (seconds = FEDFS_DELAY_MIN_SECS;; seconds = fedfs_delay(seconds)) {
		status = fedfs_get_nsdb_params_call(hostname, nettype,
						nsdbname, nsdbport, certfile);
		if (status != FEDFS_ERR_DELAY)
			break;

		xlog(D_GENERAL, "Delaying %u seconds...", seconds);
		if (sleep(seconds) != 0)
			break;
	}

	nsdb_connsec_crypto_shutdown();
	return (int)status;
}
